Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -357,92 +368,131 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
             return 0
         }
     }
 
-    public func decoderRemoved(ctx: ChannelHandlerContext) {
-        // Remove the stored reference to ChannelHandlerContext
-        parser.data = UnsafeMutableRawPointer(bitPattern: 0x0000deadbeef0000)
-
-        // Set the callbacks to nil as we dont need these anymore
-        settings.on_body = nil
-        settings.on_chunk_complete = nil
-        settings.on_url = nil
-        settings.on_status = nil
-        settings.on_chunk_header = nil
-        settings.on_chunk_complete = nil
-        settings.on_header_field = nil
-        settings.on_header_value = nil
-        settings.on_message_begin = nil
+    private func rethrowParserError() throws {
+        // Rethrow any error
+        if let error = self.state.currentError {
+            throw error
+        }
+
+        let httpError = self.parser.http_errno
+        if httpError != 0 {
+            self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
+            throw self.state.currentError!
+        }
     }
 
-    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
-        if let slice = state.slice {
-            // If we stored a slice before we need to ensure we move the readerIndex so we don't try to parse the data again. We
-            // also need to update the reader index to whatever it is now.
-            state.slice = (buffer.readerIndex, slice.length)
-            buffer.moveReaderIndex(forwardBy: state.readerIndexAdjustment)
+    // Decode HTTP until there is nothing more to decode.
+    private func decodeHTTP(ctx: ChannelHandlerContext) throws {
+        // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+        while let bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0 {
+            // Using withVeryUnsafeBytes here as this simplifies the calculation of the readerIndex which is relative to the baseAddress.
+            let result = bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
+                self.state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+                defer {
+                    self.state.baseAddress = nil
+                }
+                return self.state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count) { p in
+                    c_nio_http_parser_execute(&self.parser, &self.settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
+                }
+            }
+            try self.rethrowParserError()
+
+            /// http_parser_execute(...) should always consume the whole buffer.
+            assert(result == bufferSlice.readableBytes)
+
+            // Update readerIndex of the cumulationBuffer itself as we will refetch it in the next loop run if needed.
+            self.cumulationBuffer?.moveReaderIndex(forwardBy: result)
```

Comment:
It's *marginally* faster to use `moveReaderIndex(to: writerIndex)` I think.

Reply:
This will mess up with the loop that handles re-entrance.

Reply:
Ah shoot you're right. Nevermind then.